docs/design/copy-tracking.md

```diff
@@ -10,17 +10,686 @@ results of rebasing across remote copies.
+|
+| L rename foo->bar
+|/
+K add foo
+```
+
+When diffing `L` to `N`, we find that all files are related. Since `bar` does
+not exist in the destination, we should find a rename target to match it with.
+We pick `baz` because it's closer in the graph than `qux` is. So the diff is:
+
+* `bar` is renamed to `baz`
+* `bar` is copied to `qux`
+
+#### Example: Copy onto deleted file
+
+```
+M copy foo->bar
+|
+L delete bar
+|
+K add foo, bar
+```
+
+When diffing from `K` to `M`, we notice that `bar` has different and unrelated
+copy IDs. We present one record saying that `bar` was deleted, and one record
+saying that `bar` was copied from `foo`.
+
+When diffing from `M` to `K`, we will instead present one record that says that
+`bar` was created, and one record that says that `bar` was merged into `foo`.
+
+### Merging
+
+When merging trees, we start by rewriting each diff to match any different names
+in the destination tree. For example, if the tree conflict is `A+(B-C)+(D-E)`,
+then we will rewrite the `(B-C)` diff and the `(D-E)` diff to the paths in `A`.
+To translate the `(B-C)` diff, we calculate renames from `C` to `A` and then we
+apply those renames to both `C` and `B`. This may result in conflicts.
+
+If a file has a conflict in the copy ID, it will appear as if doesn't exist when
+materialized. It will therefore not show up the working copy until the user has
+resolved the conflict.
+
+Should we propagate changes to copies? For example, if you've modified file
+`foo` and then rebase it onto a commit that copied `foo` to `bar`, should we
+apply your change to `bar` too? Mercurial does that but Git doesn't. It's
+particularly useful when a file has been split in two. For example, let's say
+you've made various changes in file `foo` and then rebase those change onto a
+commit that split `foo` into `foo1` and `foo2` (or `foo` and `bar`). If we
+propagate the changes to both files, each change will apply successfully in one
+file (assuming the changes do not overlap with the split boundary). Each change
+will have a modify/delete conflicts in the other file. Those can relatively
+easily be resolved in favor of the deleted hunk. If we do not propagate changes,
+then changes that belong in one of the files will instead only appear as
+modify/delete conflicts in the first file and you will have to manually copy
+over the changes to the copied file.
+
+Propagating changes to copies means that rebasing a commit and then rebasing it
+back is no longer a no-op even when ignoring the "same-change rule". For
+example, if your commit modifies file `foo` and you rebase that commit onto a
+commit that copied `foo` to `bar`, and then you rebase it back, the same change
+will be applied twice to `foo`. However, thanks to the same-change rule, we
+won't consider it a conflict, so maybe it actually works well in practice.
+
+For example:
+```
+M set foo="bye"
+|
+| L rename foo->bar
+|/
+K add foo="hello"
+```
+
+When rebasing `L` onto `M`,
```

Comment:
It looks like you did not finish this sentence: When rebasing `L` onto `M`, .... ?